[stable/gangway] Gangway extra volumes and volumeMounts, document SSL/TLS w/ examples

figaw opened the pull request:
<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->

#### What this PR does / why we need it:

- Add option for mounting extra volumes onto the Gangway deployment. Trusted CA Path can now point to a file, mounted in from a secret. 
- Document the "certFile" and "keyFile" parameters of the Gangway config; and it's possible to provide these from an existing secret as well.
- Add examples and documentation for use of certData/keyData from `values.yaml` and certFile/keyFile from a secret.

#### Which issue this PR fixes

  - fixes #15521 

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

figaw committed:
ADD option for specifying extra volumes and volumeMounts #15521

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

figaw committed:
ADD documentation about certFile and keyFile from Gangway, ADD SSL configuration examples to documentation.

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

figaw committed:
Bump minor-version of chart, after adding volume, volumeMount feature

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

rk295 commented on stable/gangway/templates/deployment.yaml:
```diff
@@ -52,6 +52,9 @@ spec:
             mountPath: /etc/gangway/tls
             readOnly: true
         {{ end }}
+{{- if ne (len .Values.extraVolumeMounts) 0 }}
```
Its more common to simply do the following rather than check the length here.
```
      {{- if .Values.extraVolumes }}
```

figaw replied:
Solved in 881b35ac0.

rk295 commented on stable/gangway/templates/deployment.yaml:
```diff
@@ -52,6 +52,9 @@ spec:
             readOnly: true
         {{ end }}
+{{- if ne (len .Values.extraVolumeMounts) 0 }}
+{{ toYaml .Values.extraVolumeMounts | trim | indent 10 }}
```
`trim` not strictly necessary here, but I guess it does no harm. Is the indent 10 right? I've not tested but it feels like thats too much.

figaw replied:
Yeah, the entire container-block is indented like,
```
      containers:
        - name: {{ .Chart.Name }}
```
and not like
```
      containers:
      - name: {{ .Chart.Name }}
```
So it's 10 instead of 8. (Same for the 8 and not 6 on the extraVolumes.)

figaw replied:
Align entire chart to use compact-style block sequences in a2759b8.

rk295 commented on stable/gangway/templates/deployment.yaml:
```diff
@@ -93,4 +96,7 @@ spec:
-        {{ end -}}
\ No newline at end of file
+        {{ end -}}
+{{- if ne (len .Values.extraVolumes) 0 }}
```
same comment about the `len` call.

figaw replied:
Solved in 881b35ac0.

rk295 commented on stable/gangway/README.md:
````diff
@@ -110,4 +114,43 @@ $ helm upgrade --install --wait my-release stable/gangway -f values.yaml
+...
+```
+
+With a secret called `my-ssl-secret`:
````
Now I've read your additions to `README.md` I understand what you are doing! Sorry I was confused there for a moment. I think adding support for an existing secret like this is somewhat confusing. I think it would be simpler and less confusing to do something like:

```
        {{- if and .Values.gangway.certData .Values.gangway.keyData }}
        - name: gangway-tls
          secret:
            secretName: {{ include "gangway.fullname" . }}-tls
        {{ else << user provides an existing secret >> }}
        - name: gangway-tls
          secret:
            secretName: << insert name of existing secret provided by the user>> 
        {{ end -}}
```

(in `deployment.yaml`)

figaw replied:
Solved in c31209466.

rk295 commented on stable/gangway/README.md:
```diff
@@ -110,4 +114,43 @@ $ helm upgrade --install --wait my-release stable/gangway -f values.yaml
+...
+data:
+  tls.crt: AB..==
+  tls.key: CD..==
```
worth highlighting here that the names of the keys (i.e. `tls.crt` and `tls.key`) are _important_ !

figaw replied:
Updated explanation entirely to reflect the "new" existingSecret in c31209466.

rk295 commented on stable/gangway/values.yaml:
```diff
@@ -107,6 +107,10 @@ gangway:
   # own in procution.
   # sessionKey:
 
+extraVolumes: []
```
Despite comments above about handling secrets another way, having support for extraVolumes is always good, so I'd keep this in!

figaw committed:
UPDATE indentation to use compact-block sequences consistently.

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

figaw committed:
simplify list-length check

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

figaw committed:
CHANGE gangway.certData and .keyData to ssl. Bump minor (see msg)

certData and keyData aren't parameters that can be configured in gangway,
  so it's confusing that they're configured under the gangway-key and
  passed in the configmap,
They should just be under "some" tls-key.

Minor is bumped since the chart is still in version < 1.0, so it
  should be considered unstable and changing.

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

figaw committed:
ADD simplified existingSecret for TLS

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

figaw committed:
ADD live/readinessProbe httpGet.scheme HTTPS support

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

figaw committed:
FIX lint comments

Signed-off-by: Nicolaj Græsholt <figaw@hotmail.com>

rk295 reviewed (commented):
sorry, yesterday got away from me. LGTM!
